src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/ExpressionLambdaRewriter.cs

```diff
@@ -923,6 +923,23 @@ private BoundExpression VisitPropertyAccess(BoundPropertyAccess node)
         {
             var receiver = node.PropertySymbol.IsStatic ? _bound.Null(ExpressionType) : Visit(node.ReceiverOpt);
             var getMethod = node.PropertySymbol.GetOwnOrInheritedGetMethod();
+
+            // COMPAT: see https://github.com/dotnet/roslyn/issues/4471
+            //         old compiler used to insert casts like this and 
+            //         there are known dependencies on this kind of tree shape.
+            //
+            //         While the casts are semantically incorrect, the conditions
+            //         under which they are observable are extremely narrow:
+            //         We would have to deal with a generic T receiver which is actually a struct
+            //         that implements a property form an interface and 
+            //         the implementation of the getter must make observable mutations to the instance.
+            //
+            //         At this point it seems more appropriate to continue adding these casts.
+            if (node.ReceiverOpt?.Type.IsTypeParameter() == true)
```

Comment:
It is a bit awkward, but as I understand it would be ambiguous otherwise. 
The codegen should optimize this into a null check on ReceiverOpt and regular branch on the result (no nullables will be really involved).